pageserver/src/remote_storage/storage_sync/download.rs

```diff
@@ -131,10 +133,17 @@ where
                             layer_desination_path.display()
                         )
                     })?;
-
-                let mut destination_file = fs::File::create(&layer_desination_path)
-                    .await
-                    .with_context(|| {
+                // we perform a rename inspired by durarble_rename from file_utils.c
+                // the differnce is that we sync parent dir once before rename.
+                // For us it is not critical if rename will be rolled back. In such case
+                // file will be deleted on startup and redownloaded.
+                // For more context about durable_rename check this email from postgres mailing list:
+                // https://www.postgresql.org/message-id/56583BDD.9060302@2ndquadrant.com
```

Comment:
```suggestion
                // Perform a rename inspired by durable_rename from file_utils.c.
                // The difference is that we sync parent dir once, before the rename.
                // For us it is not critical if the rename is rolled back. In such case,
                // the file will be deleted on startup and re-downloaded.
                // For more context about durable_rename, see this thread on pgsql-hackers:
                // https://www.postgresql.org/message-id/56583BDD.9060302@2ndquadrant.com
```

Reply:
If it's OK to lose the file, why do we bother fsyncing the directory at all? It seems pretty pointless to fsync it before the rename.

Reply:
Hmmm, you're right. Given the need for metadata, sync @SomeoneToIgnore pointed to above, I think proper sync after rename is needed here. For some time, I felt that we could avoid fsync for the parent directory because we can redownload the file, but if we don't fsync the dir, rename may rollback, but metadata update will be performed to match that layer. I shouldn't cause any harm because we'll still redownload the file and overwrite metadata after that again. But it is a bit scary to have metadata with data that does not match files available on disk. What do you think?

I had a quick chat with @antons-antons touching that so would be glad to get some feedback

Reply:
No objections to fsyncing the directory too. We can revisit later if it becomes a performance problem, but better to err on the safe side.

So it would be:
1. fsync the file with old name
2. rename
3. fsync the file with new name
4. fsync parent directory

Right?

Reply:
From the description [here](https://lwn.net/Articles/457667/) the sequence is that: 
1. create a new temp file (on the same file system!)
2. write data to the temp file
3. fsync() the temp file
4. rename the temp file to the appropriate name
5. fsync() the containing directory

So the only difference with your proposal is the absence of fsync on a file after rename. Any downsides to that approach?

Reply:
The [thread on pgsql-hackers](https://www.postgresql.org/message-id/flat/56583BDD.9060302%402ndquadrant.com) concluded with the fsync() on the file, too. I'm not sure why, I don't see that point being mentioned explicitly, everyone seems to just assume that it's needed.

Based on the LWN article I think you're right that it's not needed. But why take the risk? An extra fsync() is very cheap when the file has already been flushed to disk. Better safe than sorry.

Reply:
Agreed, I've changed the code to match. Thanks!